File: frontend/src/types/detection.ts
// Types mirroring the backend Pydantic schemas in backend/app/schemas/detection.py.
// Kept in sync manually since this project does not generate a client from
// the OpenAPI schema, to keep the frontend build independent of the backend.

export interface MediaAsset {
  id: string;
  media_type: "image" | "video";
  original_filename: string;
  content_type: string | null;
  size_bytes: number | null;
  width: number | null;
  height: number | null;
  duration_seconds: number | null;
  uploaded_at: string;
}

export interface Detection {
  id: string;
  media_asset_id: string;
  frame_index: number;
  frame_timestamp: number | null;
  class_name: string;
  class_id: number;
  confidence: number;
  bbox_x_min: number;
  bbox_y_min: number;
  bbox_x_max: number;
  bbox_y_max: number;
  is_violation: boolean;
  model_name: string;
  model_version: string;
  created_at: string;
}

export interface DetectionListResponse {
  total: number;
  limit: number;
  offset: number;
  items: Detection[];
}

export interface InferenceResponse {
  job_id: string;
  media_asset: MediaAsset;
  status: string;
  total_frames: number;
  total_detections: number;
  total_violations: number;
  processing_time_ms: number;
  detections: Detection[];
}

export interface ClassSummary {
  class_name: string;
  total_detections: number;
  total_violations: number;
  avg_confidence: number | null;
  last_seen_at: string | null;
}

export interface StatsSummaryResponse {
  total_media_assets: number;
  total_detections: number;
  total_violations: number;
  compliance_rate: number;
  by_class: ClassSummary[];
}

export interface TimeseriesPoint {
  bucket_start: string;
  total_detections: number;
  total_violations: number;
}

export interface StatsTimeseriesResponse {
  interval: string;
  points: TimeseriesPoint[];
}

export interface DetectionFilters {
  class_name?: string;
  is_violation?: boolean;
  limit?: number;
  offset?: number;
}

File: backend/app/schemas/detection.py
"""Pydantic schemas for request validation and API responses."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class BoundingBox(BaseModel):
    x_min: float
    y_min: float
    x_max: float
    y_max: float


class DetectionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    media_asset_id: str
    frame_index: int
    frame_timestamp: float | None
    class_name: str
    class_id: int
    confidence: float
    bbox_x_min: float
    bbox_y_min: float
    bbox_x_max: float
    bbox_y_max: float
    is_violation: bool
    model_name: str
    model_version: str
    created_at: datetime


class DetectionListResponse(BaseModel):
    total: int
    limit: int
    offset: int
    items: list[DetectionOut]


class MediaAssetOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    media_type: str
    original_filename: str
    content_type: str | None
    size_bytes: int | None
    width: int | None
    height: int | None
    duration_seconds: float | None
    uploaded_at: datetime


class InferenceResponse(BaseModel):
    job_id: str
    media_asset: MediaAssetOut
    status: str
    total_frames: int
    total_detections: int
    total_violations: int
    processing_time_ms: int
    detections: list[DetectionOut]


class ClassSummary(BaseModel):
    class_name: str
    total_detections: int
    total_violations: int
    avg_confidence: float | None
    last_seen_at: datetime | None


class StatsSummaryResponse(BaseModel):
    total_media_assets: int
    total_detections: int
    total_violations: int
    compliance_rate: float = Field(..., description="Share of detections that are not a violation, between 0 and 1.")
    by_class: list[ClassSummary]


class TimeseriesPoint(BaseModel):
    bucket_start: datetime
    total_detections: int
    total_violations: int


class StatsTimeseriesResponse(BaseModel):
    interval: str
    points: list[TimeseriesPoint]


class ErrorResponse(BaseModel):
    detail: str
